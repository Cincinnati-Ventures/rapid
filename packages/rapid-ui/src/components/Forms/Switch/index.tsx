export { default as Switch, switchTheme } from './Switch';
