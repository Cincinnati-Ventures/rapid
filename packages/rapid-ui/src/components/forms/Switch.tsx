import React from 'react';
import { Switch as HeadlessSwitch, SwitchGroupProps } from '@headlessui/react';
import { RapidStyles } from '../../utils';
import classes from '../../conditional';
import { createVariant } from '../../theme';

const RAPID_CLASSNAME = 'rapid-switch';

interface SwitchProps extends SwitchGroupProps<typeof HeadlessSwitch> {
  styles?: string;
  ariaLabel?: string;
  enabled: boolean;
};

const Switch = React.forwardRef<
	React.ElementRef<typeof HeadlessSwitch>,
	SwitchProps
>(({ styles, ariaLabel, enabled, ...rest }, ref) => {

  const defaultStyles = 'relative inline-flex h-[28px] w-[52px] shrink-0 cursor-pointer rounded-full border-2 border-transparent transition-colors duration-200 ease-in-out focus:outline-none focus-visible:ring-2 focus-visible:ring-white focus-visible:ring-opacity-75';

  return (
      <HeadlessSwitch
        checked={enabled}
        ref={ref}
        {...rest}
        className={RapidStyles(styles || rest.className, classes(enabled ? 'bg-black' : 'bg-lightGrey', defaultStyles), RAPID_CLASSNAME)}
      >
        <span className="sr-only">{ariaLabel || 'toggle'}</span>
        <span
          aria-hidden="true"
          className={classes(enabled ? 'translate-x-6' : 'translate-x-0', ' pointer-events-none inline-block h-[24px] w-[24px] transform rounded-full bg-white shadow-lg ring-0 transition duration-200 ease-in-out')}
        />
      </HeadlessSwitch>
  )
});

export default Switch;


