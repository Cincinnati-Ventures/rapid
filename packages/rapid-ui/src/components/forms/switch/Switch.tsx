import React from 'react';
import { Switch as HeadlessSwitch, SwitchGroupProps } from '@headlessui/react';
import { RapidStyles, getVariantClassName } from '../../../utils';
import classes from '../../../conditional';
import { createVariant } from '../../../theme';
import { motion } from 'framer-motion';

const THEME_CLASSNAME = 'rapid-switch';

export const switchTheme = createVariant({
	variants: {
		default:
			'relative inline-flex shrink-0 cursor-pointer rounded-full border-2 border-transparent transition-colors duration-200 ease-in-out focus:outline-none focus-visible:ring-2 focus-visible:ring-white focus-visible:ring-opacity-75',
	},
	sizes: {
		sm: 'h-[20px] w-[40px]',
		md: 'h-[28px] w-[52px]',
		lg: 'h-[36px] w-[64px]',
	},
	defaultProps: {
		variant: 'default',
		size: 'md',
	},
});

interface SwitchProps extends SwitchGroupProps<typeof HeadlessSwitch> {
	styles?: string;
	ariaLabel?: string;
	enabled: boolean;
	size?: 'sm' | 'md' | 'lg';
	variant?: string;
	enabledStyles?: string;
	disabledStyles?: string;
}

const Switch = React.forwardRef<
	React.ElementRef<typeof HeadlessSwitch>,
	SwitchProps
>(
	(
		{
			styles,
			ariaLabel,
			enabled,
			size,
			variant,
			enabledStyles,
			disabledStyles,
			...rest
		},
		ref,
	) => {
		return (
			<HeadlessSwitch
				checked={enabled}
				ref={ref}
				{...rest}
				className={RapidStyles(
					classes(
						enabled
							? enabledStyles || 'bg-black'
							: disabledStyles || 'bg-lightGrey',
						'inline-flex ring-0 focus:ring-0',
						enabled ? 'justify-end' : 'justify-start',
					),
					styles || rest.className,
					getVariantClassName(variant, 'switch', size) ||
						THEME_CLASSNAME,
				)}
			>
				<span className='sr-only'>{ariaLabel || 'toggle'}</span>
				<motion.span
					layout
					transition={{
						duration: 0.2,
					}}
					aria-hidden='true'
					className={classes(
						'pointer-events-none inline-block aspect-square h-[100%] transform rounded-full bg-white shadow-lg ring-0 focus:ring-0',
					)}
				/>
			</HeadlessSwitch>
		);
	},
);

Switch.displayName = 'Switch';

export default Switch;
