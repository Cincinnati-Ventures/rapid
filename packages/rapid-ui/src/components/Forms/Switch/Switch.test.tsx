import React from 'react';
import { render } from '@testing-library/react';
import { expect, it, describe } from 'vitest';
import Switch from './Switch';

describe('Switch', () => {
	it('should render default correctly', () => {
		const { container } = render(<Switch enabled={true} />);
		expect(container).toMatchSnapshot();
	});

	it('should render correctly with non default props', () => {
		const { container } = render(
			<Switch
				styles='p-4'
				ariaLabel='Some label'
				enabled={false}
				size='sm'
				variant='this is an undeclared variant name'
				enabledStyles='bg-red-500'
				disabledStyles='bg-blue-500'
			/>,
		);
		expect(container).toMatchSnapshot();
	});
});
