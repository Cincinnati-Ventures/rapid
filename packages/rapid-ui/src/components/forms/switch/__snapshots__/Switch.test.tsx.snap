// Vitest Snapshot v1

exports[`Switch > should render correctly with non default props 1`] = `
<div>
  <button
    aria-checked="false"
    class="p-4 bg-blue-500 inline-flex ring-0 focus:ring-0 justify-start rapid-switch-this is an undeclared variant name rapid-switch-sm"
    data-headlessui-state=""
    id="headlessui-switch-:r1:"
    role="switch"
    tabindex="0"
    type="button"
  >
    <span
      class="sr-only"
    >
      Some label
    </span>
    <span
      aria-hidden="true"
      class="pointer-events-none inline-block aspect-square h-[100%] transform rounded-full bg-white shadow-lg ring-0 focus:ring-0"
      data-projection-id="1"
    />
  </button>
</div>
`;

exports[`Switch > should render default correctly 1`] = `
<div>
  <button
    aria-checked="true"
    class="bg-black inline-flex ring-0 focus:ring-0 justify-end rapid-switch"
    data-headlessui-state="checked"
    id="headlessui-switch-:r0:"
    role="switch"
    tabindex="0"
    type="button"
  >
    <span
      class="sr-only"
    >
      toggle
    </span>
    <span
      aria-hidden="true"
      class="pointer-events-none inline-block aspect-square h-[100%] transform rounded-full bg-white shadow-lg ring-0 focus:ring-0"
    />
  </button>
</div>
`;
